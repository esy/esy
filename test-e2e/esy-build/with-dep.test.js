// @flow

const path = require('path');
const helpers = require('../test/helpers');
const {test, isWindows, isMacos, isLinux} = helpers;

function makeFixture(p, buildDep) {
  return [
    helpers.packageJson({
      name: 'withDep',
      version: '1.0.0',
      esy: {
        build: 'true',
      },
      dependencies: {
        dep: 'path:./dep',
      },
    }),
    helpers.dir(
      'dep',
      helpers.packageJson({
        name: 'dep',
        version: '1.0.0',
        esy: {
          ...buildDep,
          exportedEnv: {
            dep__local: {val: 'dep__local'},
            dep__global: {val: 'dep__global', scope: 'global'},
            dep__local_dyn: {val: 'dep__local_dyn:$cur__name'},
            dep__global_dyn: {val: 'dep__global_dyn:$cur__name', scope: 'global'},
          },
        },
        dependencies: {
          depOfDep: 'path:../depOfDep',
        },
      }),
      helpers.dummyExecutable('dep'),
    ),
    helpers.dir(
      'depOfDep',
      helpers.packageJson({
        name: 'depOfDep',
        version: '1.0.0',
        esy: {
          build: 'true',
          exportedEnv: {
            depOfDep__local: {val: 'depOfDep__local'},
            depOfDep__global: {val: 'depOfDep__global', scope: 'global'},
            depOfDep__local_dyn: {val: 'depOfDep__local_dyn:$cur__name'},
            depOfDep__global_dyn: {
              val: 'depOfDep__global_dyn:$cur__name',
              scope: 'global',
            },
          },
        },
      }),
      helpers.dummyExecutable('dep'),
    ),
  ];
}

describe('Build with dep', () => {
  let winsysDir =
    process.platform === 'win32' ? [helpers.getWindowsSystemDirectory()] : [];

  describe('out of source build', () => {
    function withProject(assertions) {
      return async () => {
        const p = await helpers.createTestSandbox();
        await p.fixture(
          ...makeFixture(p, {
            build: [
              'cp #{self.root / self.name}.js #{self.target_dir / self.name}.js',
              helpers.buildCommand(p, '#{self.target_dir / self.name}.js'),
            ],
            install: [
              `cp #{self.target_dir / self.name}.cmd #{self.bin / self.name}.cmd`,
              `cp #{self.target_dir / self.name}.js #{self.bin / self.name}.js`,
            ],
          }),
        );
        await p.esy('install');
        await p.esy('build');
        await assertions(p);
      };
    }

    it(
      'makes dep available in envs',
      withProject(async p => {
        {
          const {stdout} = await p.esy('dep.cmd');
          expect(stdout.trim()).toEqual('__dep__');
        }

        {
          const {stdout} = await p.esy('b dep.cmd');
          expect(stdout.trim()).toEqual('__dep__');
        }

        {
          const {stdout} = await p.esy('x dep.cmd');
          expect(stdout.trim()).toEqual('__dep__');
        }
      }),
    );

    test(
      'build-env',
      withProject(async function(p) {
        const id = JSON.parse((await p.esy('build-plan')).stdout).id;
        const depId = JSON.parse((await p.esy('build-plan -p dep')).stdout).id;
        const depOfDepId = JSON.parse((await p.esy('build-plan -p depOfDep')).stdout).id;

        const {stdout} = await p.esy('build-env --json');
        const env = JSON.parse(stdout);
        expect(env).toMatchObject({
          // exports
          dep__local: 'dep__local',
          dep__global: 'dep__global',
          dep__local_dyn: 'dep__local_dyn:withDep',
          dep__global_dyn: 'dep__global_dyn:withDep',
          depOfDep__global: 'depOfDep__global',
          depOfDep__global_dyn: 'depOfDep__global_dyn:withDep',
          //built ins (cur)
          cur__version: '1.0.0',
          cur__toplevel: `${p.projectPath}/_esy/default/store/i/${id}/toplevel`,
          cur__target_dir: `${p.projectPath}/_esy/default/store/b/${id}`,
          cur__stublibs: `${p.projectPath}/_esy/default/store/i/${id}/stublibs`,
          cur__share: `${p.projectPath}/_esy/default/store/i/${id}/share`,
          cur__sbin: `${p.projectPath}/_esy/default/store/i/${id}/sbin`,
          cur__root: `${p.projectPath}`,
          cur__original_root: `${p.projectPath}`,
          cur__name: `withDep`,
          cur__man: `${p.projectPath}/_esy/default/store/i/${id}/man`,
          cur__lib: `${p.projectPath}/_esy/default/store/i/${id}/lib`,
          cur__install: `${p.projectPath}/_esy/default/store/i/${id}`,
          cur__etc: `${p.projectPath}/_esy/default/store/i/${id}/etc`,
          cur__doc: `${p.projectPath}/_esy/default/store/i/${id}/doc`,
          cur__bin: `${p.projectPath}/_esy/default/store/i/${id}/bin`,
          // built ins
          PATH: [
            `${p.esyStorePath}/i/${depId}/bin`,
            `${p.esyStorePath}/i/${depOfDepId}/bin`,
            ``,
            `/usr/local/bin`,
            `/usr/bin`,
            `/bin`,
            `/usr/sbin`,
            `/sbin`,
            ...winsysDir,
          ].join(path.delimiter),
          OCAMLFIND_LDCONF: `ignore`,
          OCAMLFIND_DESTDIR: `${p.projectPath}/_esy/default/store/i/${id}/lib`,
          DUNE_BUILD_DIR: `${p.projectPath}/_esy/default/store/b/${id}`,
        });
      }),
    );

    test.enableIf(isMacos || isLinux)(
      'macos || linux: build-env snapshot',
      withProject(async function(p) {
        const id = JSON.parse((await p.esy('build-plan')).stdout).id;
        const depId = JSON.parse((await p.esy('build-plan -p dep')).stdout).id;
        const depOfDepId = JSON.parse((await p.esy('build-plan -p depOfDep')).stdout).id;
        const {stdout} = await p.esy('build-env --build-concurrency 16');
        expect(
          p.normalizePathsForSnapshot(stdout, {id, depId, depOfDepId}),
        ).toMatchSnapshot();
      }),
    );

    test(
      'build-env dep',
      withProject(async function(p) {
        const id = JSON.parse((await p.esy('build-plan')).stdout).id;
        const depId = JSON.parse((await p.esy('build-plan -p dep')).stdout).id;
        const depOfDepId = JSON.parse((await p.esy('build-plan -p depOfDep')).stdout).id;

        const {stdout} = await p.esy('build-env --json -p dep');
        const env = JSON.parse(stdout);
        expect(env).toMatchObject({
          cur__version: '1.0.0',
          cur__toplevel: `${p.esyStorePath}/s/${depId}/toplevel`,
          cur__target_dir: `${p.esyPrefixPath}/3/b/${depId}`,
          cur__stublibs: `${p.esyStorePath}/s/${depId}/stublibs`,
          cur__share: `${p.esyStorePath}/s/${depId}/share`,
          cur__sbin: `${p.esyStorePath}/s/${depId}/sbin`,
          cur__name: `dep`,
          cur__man: `${p.esyStorePath}/s/${depId}/man`,
          cur__lib: `${p.esyStorePath}/s/${depId}/lib`,
          cur__install: `${p.esyStorePath}/s/${depId}`,
          cur__etc: `${p.esyStorePath}/s/${depId}/etc`,
          cur__doc: `${p.esyStorePath}/s/${depId}/doc`,
          cur__bin: `${p.esyStorePath}/s/${depId}/bin`,
          PATH: [
            `${p.esyStorePath}/i/${depOfDepId}/bin`,
            ``,
            `/usr/local/bin`,
            `/usr/bin`,
            `/bin`,
            `/usr/sbin`,
            `/sbin`,
            ...winsysDir,
          ].join(path.delimiter),
          OCAMLFIND_LDCONF: `ignore`,
          OCAMLFIND_DESTDIR: `${p.esyStorePath}/s/${depId}/lib`,
          DUNE_BUILD_DIR: `${p.esyPrefixPath}/3/b/${depId}`,
        });
      }),
    );

    test.enableIf(isMacos || isLinux)(
      'macos || linux: build-env dep snapshot',
      withProject(async function(p) {
        const id = JSON.parse((await p.esy('build-plan -p dep')).stdout).id;
        const depOfDepId = JSON.parse((await p.esy('build-plan -p depOfDep')).stdout).id;
        const {stdout} = await p.esy('build-env -p dep --build-concurrency 16');
        expect(p.normalizePathsForSnapshot(stdout, {id, depOfDepId})).toMatchSnapshot();
      }),
    );

    test(
      'exec-env',
      withProject(async function(p) {
        const id = JSON.parse((await p.esy('build-plan')).stdout).id;
        const depId = JSON.parse((await p.esy('build-plan -p dep')).stdout).id;
        const {stdout} = await p.esy('exec-env --json');
        const envpath = JSON.parse(stdout).PATH.split(path.delimiter);
        expect(
          envpath.includes(`${p.projectPath}/_esy/default/store/i/${id}/bin`),
        ).toBeTruthy();
        expect(envpath.includes(`${p.esyStorePath}/i/${depId}/bin`)).toBeTruthy();
      }),
    );

    test(
      'command-env',
      withProject(async function(p) {
        const id = JSON.parse((await p.esy('build-plan')).stdout).id;
        const depId = JSON.parse((await p.esy('build-plan -p dep')).stdout).id;
        const depOfDepId = JSON.parse((await p.esy('build-plan -p depOfDep')).stdout).id;
        const {stdout} = await p.esy('command-env --json');
        const env = JSON.parse(stdout);
        expect(env).toMatchObject({
          cur__version: '1.0.0',
          cur__toplevel: `${p.projectPath}/_esy/default/store/i/${id}/toplevel`,
          cur__target_dir: `${p.projectPath}/_esy/default/store/b/${id}`,
          cur__stublibs: `${p.projectPath}/_esy/default/store/i/${id}/stublibs`,
          cur__share: `${p.projectPath}/_esy/default/store/i/${id}/share`,
          cur__sbin: `${p.projectPath}/_esy/default/store/i/${id}/sbin`,
          cur__root: `${p.projectPath}`,
          cur__original_root: `${p.projectPath}`,
          cur__name: `withDep`,
          cur__man: `${p.projectPath}/_esy/default/store/i/${id}/man`,
          cur__lib: `${p.projectPath}/_esy/default/store/i/${id}/lib`,
          cur__install: `${p.projectPath}/_esy/default/store/i/${id}`,
          cur__etc: `${p.projectPath}/_esy/default/store/i/${id}/etc`,
          cur__doc: `${p.projectPath}/_esy/default/store/i/${id}/doc`,
          cur__bin: `${p.projectPath}/_esy/default/store/i/${id}/bin`,
          OCAMLFIND_LDCONF: `ignore`,
          OCAMLFIND_DESTDIR: `${p.projectPath}/_esy/default/store/i/${id}/lib`,
          DUNE_BUILD_DIR: `${p.projectPath}/_esy/default/store/b/${id}`,
        });
        const envpath = env.PATH.split(path.delimiter);
        expect(envpath.includes(`${p.esyStorePath}/i/${depId}/bin`)).toBeTruthy();
        expect(envpath.includes(`${p.esyStorePath}/i/${depOfDepId}/bin`)).toBeTruthy();
      }),
    );
  });

  describe('in source build', () => {
    function withProject(assertions) {
      return async () => {
        const p = await helpers.createTestSandbox();
        await p.fixture(
          ...makeFixture(p, {
            buildsInSource: true,
            build: [helpers.buildCommand(p, '#{self.root / self.name}.js')],
            install: [
              `cp #{self.root / self.name}.cmd #{self.bin / self.name}.cmd`,
              `cp #{self.root / self.name}.js #{self.bin / self.name}.js`,
            ],
          }),
        );
        await p.esy('install');
        await p.esy('build');
        await assertions(p);
      };
    }

    it(
      'makes dep available in envs',
      withProject(async p => {
        {
          const {stdout} = await p.esy('dep.cmd');
          expect(stdout.trim()).toEqual('__dep__');
        }

        {
          const {stdout} = await p.esy('b dep.cmd');
          expect(stdout.trim()).toEqual('__dep__');
        }

        {
          const {stdout} = await p.esy('x dep.cmd');
          expect(stdout.trim()).toEqual('__dep__');
        }
      }),
    );
  });

  describe('_build build', () => {
    function withProject(assertions) {
      return async () => {
        const p = await helpers.createTestSandbox();
        await p.fixture(
          ...makeFixture(p, {
            buildsInSource: '_build',
            build: [
              "mkdir -p #{self.root / '_build'}",
              "cp #{self.root / self.name}.js #{self.root / '_build' / self.name}.js",
              helpers.buildCommand(p, "#{self.root / '_build' / self.name}.js"),
            ],
            install: [
              `cp #{self.root / '_build' / self.name}.cmd #{self.bin / self.name}.cmd`,
              `cp #{self.root / '_build' / self.name}.js #{self.bin / self.name}.js`,
            ],
          }),
        );
        await p.esy('install');
        await p.esy('build');
        await assertions(p);
      };
    }
    it(
      'makes dep available in envs',
      withProject(async p => {
        {
          const {stdout} = await p.esy('dep.cmd');
          expect(stdout.trim()).toEqual('__dep__');
        }

        {
          const {stdout} = await p.esy('b dep.cmd');
          expect(stdout.trim()).toEqual('__dep__');
        }

        {
          const {stdout} = await p.esy('x dep.cmd');
          expect(stdout.trim()).toEqual('__dep__');
        }
      }),
    );
  });

  describe('out of source build with buildDev in deps', () => {
    function withProject(assertions) {
      return async () => {
        const p = await helpers.createTestSandbox();
        await p.fixture(
          ...makeFixture(p, {
            build: [
              'cp #{self.root / self.name}.js #{self.target_dir / self.name}.js',
              helpers.buildCommand(p, '#{self.target_dir / self.name}.js'),
            ],
            // set buildDev to false to make sure we fail if it's going to be
            // executed
            buildDev: 'false',
            install: [
              `cp #{self.target_dir / self.name}.cmd #{self.bin / self.name}.cmd`,
              `cp #{self.target_dir / self.name}.js #{self.bin / self.name}.js`,
            ],
          }),
        );
        await p.esy('install');
        await p.esy('build');
        await assertions(p);
      };
    }

    it(
      'makes dep available in envs',
      withProject(async p => {
        {
          const {stdout} = await p.esy('dep.cmd');
          expect(stdout.trim()).toEqual('__dep__');
        }

        {
          const {stdout} = await p.esy('b dep.cmd');
          expect(stdout.trim()).toEqual('__dep__');
        }

        {
          const {stdout} = await p.esy('x dep.cmd');
          expect(stdout.trim()).toEqual('__dep__');
        }
      }),
    );
  });
});
