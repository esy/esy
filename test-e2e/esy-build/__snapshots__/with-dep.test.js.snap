// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Build with dep out of source build macos || linux: build-env dep snapshot 1`] = `
"# Build environment
# package:            dep@path:dep
# depspec:            dependencies(self)
# mode:               build
# envspec:            
# buildIsInProgress:  true
# includeBuildEnv:    true
# includeCurrentEnv:  false
# includeNpmBin:      false


#
# Built-in
#
export cur__jobs="8"
export cur__etc="%esyStorePath%/s/%id%/etc"
export cur__share="%esyStorePath%/s/%id%/share"
export cur__toplevel="%esyStorePath%/s/%id%/toplevel"
export cur__stublibs="%esyStorePath%/s/%id%/stublibs"
export cur__doc="%esyStorePath%/s/%id%/doc"
export cur__man="%esyStorePath%/s/%id%/man"
export cur__lib="%esyStorePath%/s/%id%/lib"
export cur__sbin="%esyStorePath%/s/%id%/sbin"
export cur__bin="%esyStorePath%/s/%id%/bin"
export cur__install="%esyStorePath%/s/%id%"
export cur__target_dir="%esyPrefixPath%/3/b/%id%"
export cur__original_root="%esyPrefixPath%/source/i/dep__5d502355"
export cur__root="%esyPrefixPath%/source/i/dep__5d502355"
export cur__dev="false"
export cur__version="1.0.0"
export cur__name="dep"

#
# depOfDep@path:depOfDep@d41d8cd9
#
export depOfDep__local="depOfDep__local"
export depOfDep__local_dyn="depOfDep__local_dyn:$cur__name"
export depOfDep__global="depOfDep__global"
export depOfDep__global_dyn="depOfDep__global_dyn:$cur__name"
export CAML_LD_LIBRARY_PATH="%esyStorePath%/i/%depOfDepId%/stublibs:%esyStorePath%/i/%depOfDepId%/lib/stublibs:$CAML_LD_LIBRARY_PATH"
export OCAMLPATH="%esyStorePath%/i/%depOfDepId%/lib:$OCAMLPATH"
export MAN_PATH="%esyStorePath%/i/%depOfDepId%/man:$MAN_PATH"
export PATH="%esyStorePath%/i/%depOfDepId%/bin:$PATH"

#
# Built-in
#
export OCAMLFIND_LDCONF="ignore"
export OCAMLFIND_DESTDIR="%esyStorePath%/s/%id%/lib"
export DUNE_BUILD_DIR="%esyPrefixPath%/3/b/%id%"
export SHELL="env -i /bin/bash --norc --noprofile"
export PATH="$PATH:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin""
`;

exports[`Build with dep out of source build macos || linux: build-env snapshot 1`] = `
"# Build environment
# package:            withDep@link-dev:./package.json
# depspec:            dependencies(self)+devDependencies(self)
# mode:               buildDev
# envspec:            
# buildIsInProgress:  true
# includeBuildEnv:    true
# includeCurrentEnv:  false
# includeNpmBin:      false


#
# Built-in
#
export cur__jobs="8"
export cur__etc="%projectPath%/_esy/default/store/i/%id%/etc"
export cur__share="%projectPath%/_esy/default/store/i/%id%/share"
export cur__toplevel="%projectPath%/_esy/default/store/i/%id%/toplevel"
export cur__stublibs="%projectPath%/_esy/default/store/i/%id%/stublibs"
export cur__doc="%projectPath%/_esy/default/store/i/%id%/doc"
export cur__man="%projectPath%/_esy/default/store/i/%id%/man"
export cur__lib="%projectPath%/_esy/default/store/i/%id%/lib"
export cur__sbin="%projectPath%/_esy/default/store/i/%id%/sbin"
export cur__bin="%projectPath%/_esy/default/store/i/%id%/bin"
export cur__install="%projectPath%/_esy/default/store/i/%id%"
export cur__target_dir="%projectPath%/_esy/default/store/b/%id%"
export cur__original_root="%projectPath%"
export cur__root="%projectPath%"
export cur__dev="true"
export cur__version="1.0.0"
export cur__name="withDep"

#
# depOfDep@path:depOfDep@d41d8cd9
#
export depOfDep__global="depOfDep__global"
export depOfDep__global_dyn="depOfDep__global_dyn:$cur__name"
export CAML_LD_LIBRARY_PATH="%esyStorePath%/i/%depOfDepId%/stublibs:%esyStorePath%/i/%depOfDepId%/lib/stublibs:$CAML_LD_LIBRARY_PATH"
export OCAMLPATH="%esyStorePath%/i/%depOfDepId%/lib:$OCAMLPATH"
export MAN_PATH="%esyStorePath%/i/%depOfDepId%/man:$MAN_PATH"
export PATH="%esyStorePath%/i/%depOfDepId%/bin:$PATH"

#
# dep@path:dep@d41d8cd9
#
export dep__local="dep__local"
export dep__local_dyn="dep__local_dyn:$cur__name"
export dep__global="dep__global"
export dep__global_dyn="dep__global_dyn:$cur__name"
export CAML_LD_LIBRARY_PATH="%esyStorePath%/i/%depId%/stublibs:%esyStorePath%/i/%depId%/lib/stublibs:$CAML_LD_LIBRARY_PATH"
export OCAMLPATH="%esyStorePath%/i/%depId%/lib:$OCAMLPATH"
export MAN_PATH="%esyStorePath%/i/%depId%/man:$MAN_PATH"
export PATH="%esyStorePath%/i/%depId%/bin:$PATH"

#
# Built-in
#
export OCAMLFIND_LDCONF="ignore"
export OCAMLFIND_DESTDIR="%projectPath%/_esy/default/store/i/%id%/lib"
export DUNE_BUILD_DIR="%projectPath%/_esy/default/store/b/%id%"
export DUNE_STORE_ORIG_SOURCE_DIR="true"
export SHELL="env -i /bin/bash --norc --noprofile"
export PATH="$PATH:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin""
`;
